File: Sources/VoxLocalCore/Models/VoxLocalError.swift
import Foundation

/// User-facing errors of the dictation pipeline. `messageKey` maps to a
/// localized string; `detail` carries technical context for the log only.
public enum VoxLocalError: Error, Equatable {
    case microphonePermissionDenied
    case microphoneUnavailable
    case recordingFailed(String)
    case emptyRecording
    case whisperBinaryMissing(String)
    case modelMissing(String)
    case modelInvalid(String)
    case transcriptionFailed(exitCode: Int32, detail: String)
    case transcriptionTimeout
    case transcriptionOutputMissing
    case unsupportedArchitecture(String)
    case hotkeyConflict(String)
    case insertionTargetGone
    case cancelled

    public var messageKey: String {
        switch self {
        case .microphonePermissionDenied: return "error.mic.denied"
        case .microphoneUnavailable: return "error.mic.unavailable"
        case .recordingFailed: return "error.recording.failed"
        case .emptyRecording: return "error.recording.empty"
        case .whisperBinaryMissing: return "error.whisper.binary"
        case .modelMissing: return "error.whisper.model.missing"
        case .modelInvalid: return "error.whisper.model.invalid"
        case .transcriptionFailed: return "error.whisper.failed"
        case .transcriptionTimeout: return "error.whisper.timeout"
        case .transcriptionOutputMissing: return "error.whisper.output"
        case .unsupportedArchitecture: return "error.whisper.arch"
        case .hotkeyConflict: return "error.hotkey.conflict"
        case .insertionTargetGone: return "error.insert.target.gone"
        case .cancelled: return "state.cancelled"
        }
    }

    /// Technical detail for logs. Never contains transcript text or audio.
    public var logDetail: String {
        switch self {
        case .recordingFailed(let d): return "recording failed: \(d)"
        case .whisperBinaryMissing(let p): return "whisper-cli not found, searched: \(d(p))"
        case .modelMissing(let p): return "model missing at \(d(p))"
        case .modelInvalid(let p): return "model invalid at \(d(p))"
        case .transcriptionFailed(let code, let detail): return "whisper exit \(code): \(detail)"
        case .unsupportedArchitecture(let a): return "unsupported architecture: \(a)"
        case .hotkeyConflict(let c): return "hotkey conflict: \(c)"
        default: return String(describing: self)
        }
    }

    private func d(_ path: String) -> String { PathRedactor.redact(path) }
}

/// Replaces the user's home directory prefix in paths before logging.
public enum PathRedactor {
    public static func redact(_ path: String) -> String {
        let home = NSHomeDirectory()
        guard path.hasPrefix(home) else { return path }
        return "~" + path.dropFirst(home.count)
    }
}

File: windows/src/VoxLocal.App/Core/Models/VoxLocalError.cs
namespace VoxLocal.Core.Models;

/// <summary>
/// User-facing errors of the dictation pipeline. <see cref="MessageKey"/> maps to a
/// localized string; detail values carry technical context for the log only.
/// </summary>
public abstract record VoxLocalError
{
    public sealed record MicrophonePermissionDenied : VoxLocalError;
    public sealed record MicrophoneUnavailable : VoxLocalError;
    public sealed record RecordingFailed(string Detail) : VoxLocalError;
    public sealed record EmptyRecording : VoxLocalError;
    public sealed record WhisperBinaryMissing(string SearchedPath) : VoxLocalError;
    public sealed record ModelMissing(string Path) : VoxLocalError;
    public sealed record ModelInvalid(string Path) : VoxLocalError;
    public sealed record TranscriptionFailed(int ExitCode, string Detail) : VoxLocalError;
    public sealed record TranscriptionTimeout : VoxLocalError;
    public sealed record TranscriptionOutputMissing : VoxLocalError;
    public sealed record UnsupportedArchitecture(string Architecture) : VoxLocalError;
    public sealed record HotkeyConflict(string Combo) : VoxLocalError;
    public sealed record InsertionTargetGone : VoxLocalError;
    public sealed record Cancelled : VoxLocalError;

    public string MessageKey => this switch
    {
        MicrophonePermissionDenied => "error.mic.denied",
        MicrophoneUnavailable => "error.mic.unavailable",
        RecordingFailed => "error.recording.failed",
        EmptyRecording => "error.recording.empty",
        WhisperBinaryMissing => "error.whisper.binary",
        ModelMissing => "error.whisper.model.missing",
        ModelInvalid => "error.whisper.model.invalid",
        TranscriptionFailed => "error.whisper.failed",
        TranscriptionTimeout => "error.whisper.timeout",
        TranscriptionOutputMissing => "error.whisper.output",
        UnsupportedArchitecture => "error.whisper.arch",
        HotkeyConflict => "error.hotkey.conflict",
        InsertionTargetGone => "error.insert.target.gone",
        Cancelled => "state.cancelled",
        _ => throw new InvalidOperationException($"unknown error case: {GetType().Name}"),
    };

    /// <summary>Technical detail for logs. Never contains transcript text or audio.</summary>
    public string LogDetail => this switch
    {
        RecordingFailed e => $"recording failed: {e.Detail}",
        WhisperBinaryMissing e => $"whisper-cli not found, searched: {PathRedactor.Redact(e.SearchedPath)}",
        ModelMissing e => $"model missing at {PathRedactor.Redact(e.Path)}",
        ModelInvalid e => $"model invalid at {PathRedactor.Redact(e.Path)}",
        TranscriptionFailed e => $"whisper exit {e.ExitCode}: {e.Detail}",
        UnsupportedArchitecture e => $"unsupported architecture: {e.Architecture}",
        HotkeyConflict e => $"hotkey conflict: {e.Combo}",
        _ => ToString(),
    };
}

/// <summary>Exception wrapper so the error union can flow through async call stacks.</summary>
public sealed class VoxLocalException : Exception
{
    public VoxLocalError Error { get; }

    public VoxLocalException(VoxLocalError error) : base(error.LogDetail)
    {
        Error = error;
    }
}

/// <summary>Replaces the user's profile directory prefix in paths before logging.</summary>
public static class PathRedactor
{
    public static string Redact(string path)
    {
        var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        if (string.IsNullOrEmpty(home) || !path.StartsWith(home, StringComparison.OrdinalIgnoreCase))
        {
            return path;
        }
        return "~" + path[home.Length..];
    }
}
